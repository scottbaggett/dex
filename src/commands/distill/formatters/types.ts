/**
 * Formatter types and interfaces
 */

import { DistillationResult, CompressionResult } from '../../../types';

export interface FormatterOptions {
    // Control what to include
    includeImports?: boolean;
    includePrivate?: boolean;
    includeDocstrings?: boolean;
    includeComments?: boolean;
    includeMetadata?: boolean;
    
    // Output style
    preserveStructure?: boolean;
    groupByType?: boolean;
    
    // Language hints
    language?: string;
    syntaxHighlight?: boolean;
}

export interface Formatter {
    name: string;
    format: string;  // 'xml', 'markdown', 'json', 'text'
    
    /**
     * Format a distillation result
     */
    formatDistillation(
        result: DistillationResult,
        options?: FormatterOptions
    ): string;
    
    /**
     * Format a compression result
     */
    formatCompression(
        result: CompressionResult,
        options?: FormatterOptions
    ): string;
    
    /**
     * Format combined results
     */
    formatCombined(
        compression: CompressionResult,
        distillation: DistillationResult,
        options?: FormatterOptions
    ): string;
}

export interface FormatterRegistry {
    register(formatter: Formatter): void;
    get(format: string): Formatter | null;
    getDefault(): Formatter;
    list(): string[];
}