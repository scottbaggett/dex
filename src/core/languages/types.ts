/**
 * Language module types and interfaces
 * Each language defines its own parser type and processing strategy
 */

import { PrivateIdentifier } from "ts-morph";

export type ParserType = "tree-sitter" | "line-based" | "regex" | "custom";

/**
 * Processing options from CLI/user
 */
export interface ProcessingOptions {
    // Output control
    comments?: boolean; // Include comment nodes
    docstrings?: boolean; // Include documentation
    public?: boolean; // Include public members
    private?: boolean; // Include private members
    protected?: boolean; // Include protected members
    internal?: boolean; // Include internal members

    // Depth control
    maxDepth?: number; // Maximum nesting depth

    // Filtering
    include?: string[]; // Include only matching names
    exclude?: string[]; // Exclude matching names

    // Output format hints
    preserveOrder?: boolean; // Maintain source order

    // Performance
    maxFileSize?: number; // Skip files larger than this
    timeout?: number; // Processing timeout in ms
}

export interface LanguageModule {
    name: string; // 'typescript', 'python', etc.
    displayName: string; // 'TypeScript', 'Python', etc.
    extensions: string[]; // ['.ts', '.tsx', '.js', '.jsx']

    // Lifecycle
    initialize(): Promise<void>;
    isInitialized(): boolean;

    // Processing - MUST respect options
    process(
        source: string,
        filePath: string,
        options: ProcessingOptions,
    ): Promise<ProcessResult>;

    // Capabilities
    getCapabilities(): LanguageCapabilities;

    // Validation
    canProcess(filePath: string): boolean;
    supportsExtension(ext: string): boolean;
}

export interface LanguageCapabilities {
    supportsPrivateMembers: boolean;
    supportsComments: boolean;
    supportsDocstrings: boolean;
    supportsStreaming: boolean;
    supportsPartialParsing: boolean;
    maxRecommendedFileSize?: number;
}

export interface ProcessResult {
    imports: ImportNode[];
    exports: ExportNode[];
    errors?: ProcessError[];
    metadata?: {
        parseTime?: number;
        nodeCount?: number;
        skipped?: SkippedItem[]; // What was filtered out
        languageVersion?: string;
    };
}

export interface SkippedItem {
    name: string;
    reason: "private" | "pattern" | "depth" | "comment";
    line?: number;
}

export interface ImportNode {
    source: string;
    specifiers: ImportSpecifier[];
    line?: number;
    raw?: string; // Original import statement
}

export interface ImportSpecifier {
    name: string;
    alias?: string;
    isDefault?: boolean;
    isNamespace?: boolean;
}

export interface ExportNode {
    name: string;
    kind: ExportKind;
    signature: string;
    visibility?: "public" | "private" | "protected";
    members?: MemberNode[];
    line?: number;
    depth?: number; // Nesting depth
    isDefault?: boolean;
    isExported?: boolean;
    docstring?: string; // If includeDocstrings is true
    comment?: string; // If includeComments is true
    raw?: string; // Original declaration
}

export type ExportKind =
    | "function"
    | "class"
    | "interface"
    | "type"
    | "enum"
    | "const"
    | "let"
    | "var"
    | "namespace"
    | "module";

export interface MemberNode {
    name: string;
    kind: "property" | "method" | "getter" | "setter" | "constructor";
    signature: string;
    isStatic?: boolean;
    isPrivate?: boolean;
    isProtected?: boolean;
    isOptional?: boolean;
}

export interface ProcessError {
    message: string;
    line?: number;
    column?: number;
    severity?: "error" | "warning" | "info";
}
