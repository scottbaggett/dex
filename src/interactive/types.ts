import type { GitChange } from "../types.js";

export interface FileItem extends GitChange {
    selected: boolean;
}

export interface InteractiveState {
    files: FileItem[];
    cursor: number;
    selectedCount: number;
    totalAdditions: number;
    totalDeletions: number;
    estimatedTokens: number;
}

export interface InteractiveModeProps {
    changes: GitChange[];
    onComplete: (selectedFiles: GitChange[], copyToClipboard?: boolean) => void;
    onCancel: () => void;
}
