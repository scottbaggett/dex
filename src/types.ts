export type DepthLevel = 'minimal' | 'focused' | 'full' | 'extended';
export type OutputFormat = 'markdown' | 'json' | 'claude' | 'gpt' | 'gemini' | 'grok' | 'llama' | 'mistral' | 'github-pr' | 'pr';

export interface DexOptions {
  // Git options
  since?: string;
  range?: string;
  staged?: boolean;
  all?: boolean;
  
  // Snapshot options
  snapshot?: string;
  isSnapshot?: boolean;
  
  // Time-based options
  timeRange?: string;  // e.g., "2h", "30m", "1d"
  isTimeRange?: boolean;
  
  // Depth options
  depth?: DepthLevel;
  fullFiles?: string[];
  bootstrap?: boolean;
  
  // Untracked files
  includeUntracked?: boolean;
  untrackedPattern?: string;
  
  // Filter options
  path?: string;
  type?: string[];
  
  // Output options
  format?: OutputFormat;
  clipboard?: boolean;
  
  // Task integration
  task?: string;          // Direct task description
  taskFile?: string;      // Path to task file
  taskUrl?: string;       // URL to fetch task from
  taskStdin?: boolean;    // Read task from stdin (--task=-)
  interactive?: boolean;
  
  // File selection
  select?: boolean;       // Interactive file selection mode
  
  // Optimization (from --optimize flag)
  symbols?: boolean;
  aid?: boolean;
  
  // Display options
  noMetadata?: boolean;
  
  // Prompt options
  prompt?: string;           // Custom prompt text
  promptTemplate?: string;   // Prompt template name (security, performance, etc.)
  noPrompt?: boolean;        // Disable prompt generation
}

export interface GitChange {
  file: string;
  status: 'added' | 'modified' | 'deleted' | 'renamed';
  additions: number;
  deletions: number;
  diff: string;
  oldFile?: string; // for renames
  lastModified?: Date; // file modification time
}

export interface Metadata {
  generated: string; // ISO timestamp
  repository: {
    name: string;
    branch: string;
    commit: string;
  };
  extraction: {
    depth: DepthLevel;
    filters?: {
      path?: string;
      type?: string[];
    };
  };
  tokens: {
    estimated: number;
  };
  tool: {
    name: string;
    version: string;
  };
}

export interface ExtractedContext {
  changes: GitChange[];
  scope: {
    filesChanged: number;
    functionsModified: number;
    linesAdded: number;
    linesDeleted: number;
  };
  task?: TaskContext;
  fullFiles?: Map<string, string>;
  symbols?: SymbolMap;
  metadata: Metadata;
}

export interface TaskContext {
  description: string;
  goals?: string[];
  issueUrl?: string;
  issueTitle?: string;
  issueBody?: string;
  labels?: string[];
}

export interface SymbolMap {
  [file: string]: {
    functions: string[];
    classes: string[];
    exports: string[];
    imports: string[];
  };
}

export interface FormatterOptions {
  context: ExtractedContext;
  options: DexOptions;
}

export interface Formatter {
  format(options: FormatterOptions): string;
}

// Snapshot-related types
export interface SnapshotMetadata {
  id: string;
  time: string; // ISO timestamp
  description?: string;
  tags?: string[];
  filesCount: number;
  totalSize: number;
}

export interface SnapshotTree {
  files: {
    [path: string]: {
      hash: string;
      size: number;
      mode: string;
    };
  };
}

export interface Snapshot {
  metadata: SnapshotMetadata;
  tree: SnapshotTree;
}

export interface SnapshotDiff {
  added: string[];
  modified: string[];
  deleted: string[];
  renamed: Array<{ from: string; to: string }>;
}

export interface SnapshotOptions {
  message?: string;
  tags?: string[];
  includeUntracked?: boolean;
  path?: string;
  ignorePatterns?: string[];
  onProgress?: (progress: { current: number; total: number; file: string }) => void;
}
