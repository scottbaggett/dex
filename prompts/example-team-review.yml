# Example custom prompt template for team-specific code review
# Place this file in your project's prompts/ directory

name: Team Code Review
description: Our team's specific code review guidelines
extends: base-review  # Inherit from built-in base-review template
tags:
  - team
  - custom
  - review
llm:
  - claude
  - gpt
instructions: |
  Perform a comprehensive code review following our team standards:

  ## 1. Architecture & Design
  - Does this follow our microservices patterns?
  - Are domain boundaries respected?
  - Is the code in the right service/module?
  
  ## 2. Our Coding Standards
  - TypeScript strict mode compliance
  - Proper error handling with our custom ErrorTypes
  - Logging using our centralized logger
  - Feature flags for new functionality
  
  ## 3. Database & Performance
  - Are database queries optimized?
  - Proper use of our caching layer
  - No N+1 query problems
  - Appropriate indexes defined
  
  ## 4. Security Requirements
  - Authentication using our OAuth2 middleware
  - Proper authorization checks
  - Input validation with our validation library
  - No PII in logs
  
  ## 5. Testing Standards
  - Unit tests with >80% coverage
  - Integration tests for API endpoints
  - E2E tests for critical user flows
  - Performance tests for heavy operations

  Format your review as:
  - ✅ **Good**: What was done well
  - ⚠️ **Concern**: Issues that need addressing
  - 💡 **Suggestion**: Improvements to consider
  - 🚫 **Blocker**: Must fix before merge

# Variables can be used for dynamic content
variables:
  team_name: Engineering Team
  min_coverage: "80"
  
# Examples help LLMs understand the expected output format
examples:
  - input: "New API endpoint without authentication"
    output: |
      🚫 **Blocker**: Missing authentication
      **Location**: api/users.ts:45
      **Issue**: The POST /users/profile endpoint lacks authentication middleware
      **Fix**: Add `authMiddleware` to the route: `router.post('/profile', authMiddleware, updateProfile)`