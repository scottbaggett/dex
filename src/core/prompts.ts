import { ExtractedContext, DexOptions } from '../types';
import { PromptLoader } from './prompt-loader';

export interface PromptContext {
  task?: string;
  filesChanged: number;
  linesAdded: number;
  linesDeleted: number;
  fileTypes: Set<string>;
  hasTests: boolean;
  hasConfig: boolean;
  hasDocs: boolean;
  primaryLanguage: string;
}

export class PromptGenerator {
  private static analyzeContext(context: ExtractedContext): PromptContext {
    const fileTypes = new Set<string>();
    let hasTests = false;
    let hasConfig = false;
    let hasDocs = false;
    
    // Analyze file types and patterns
    for (const change of context.changes) {
      // Handle both file and path properties for compatibility
      const filePath = (change as any).path || change.file;
      const ext = filePath.split('.').pop() || '';
      fileTypes.add(ext);
      
      const pathLower = filePath.toLowerCase();
      if (pathLower.includes('test') || pathLower.includes('spec')) hasTests = true;
      if (pathLower.includes('config') || pathLower.includes('.env') || pathLower.includes('settings')) hasConfig = true;
      if (pathLower.includes('readme') || pathLower.includes('docs') || ext === 'md') hasDocs = true;
    }
    
    // Determine primary language
    const langPriority = ['ts', 'tsx', 'js', 'jsx', 'py', 'go', 'rs', 'java', 'cpp', 'c'];
    let primaryLanguage = 'unknown';
    for (const lang of langPriority) {
      if (fileTypes.has(lang)) {
        primaryLanguage = lang;
        break;
      }
    }
    
    return {
      task: context.task?.description,
      filesChanged: context.scope.filesChanged,
      linesAdded: context.scope.linesAdded,
      linesDeleted: context.scope.linesDeleted,
      fileTypes,
      hasTests,
      hasConfig,
      hasDocs,
      primaryLanguage
    };
  }
  
  static generate(context: ExtractedContext, options: DexOptions): string {
    const analysis = this.analyzeContext(context);
    const promptLoader = PromptLoader.getInstance();
    
    // If custom prompt provided, use it
    if (options.prompt) {
      return options.prompt;
    }
    
    // If prompt template requested, use it
    if (options.promptTemplate) {
      const template = promptLoader.getPrompt(options.promptTemplate);
      if (template) {
        let instructions = template.instructions;
        
        // Interpolate variables if provided
        if (template.variables) {
          const variables = {
            ...template.variables,
            task: analysis.task || '',
            file_count: String(analysis.filesChanged),
            primary_language: analysis.primaryLanguage
          };
          instructions = promptLoader.interpolateVariables(instructions, variables);
        }
        
        return instructions;
      }
      
      // Fall back to old system if template not found
      console.warn(`Prompt template '${options.promptTemplate}' not found, using default`);
      return this.getPromptByName(options.promptTemplate, analysis);
    }
    
    // Auto-suggest prompts if in interactive mode
    if (options.interactive && !options.noPrompt) {
      const suggestions = promptLoader.suggestPrompts({
        format: options.format,
        task: analysis.task,
        fileTypes: analysis.fileTypes
      });
      
      if (suggestions.length > 0) {
        // In a real implementation, this would be shown in interactive UI
        // For now, we'll just use the top suggestion
        const topSuggestion = suggestions[0];
        console.log(`ðŸ’¡ Suggested prompt template: ${topSuggestion.name}`);
      }
    }
    
    // Generate contextual prompt based on what changed
    return this.generateContextualPrompt(analysis, options);
  }
  
  private static generateContextualPrompt(analysis: PromptContext, options: DexOptions): string {
    const parts: string[] = [];
    
    // If task provided, make it the primary focus
    if (analysis.task) {
      parts.push(`Review these changes in the context of: "${analysis.task}"`);
      parts.push('');
      parts.push('Please analyze:');
      parts.push('1. Whether the changes properly address the task');
      parts.push('2. Any missing implementations or edge cases');
      parts.push('3. Potential side effects or breaking changes');
    } else {
      // No task - infer intent from changes
      const changeSize = analysis.linesAdded + analysis.linesDeleted;
      
      if (changeSize < 50) {
        parts.push('Review this focused change for:');
      } else if (changeSize < 200) {
        parts.push('Review these changes for:');
      } else {
        parts.push('Review this substantial changeset for:');
      }
      
      parts.push('');
      
      // Contextual analysis points
      if (analysis.hasTests) {
        parts.push('- Test coverage and quality');
        parts.push('- Whether tests properly validate the implementation');
      }
      
      if (analysis.hasConfig) {
        parts.push('- Configuration changes and their impact');
        parts.push('- Environment compatibility');
      }
      
      if (analysis.hasDocs) {
        parts.push('- Documentation accuracy and completeness');
        parts.push('- Whether docs match the implementation');
      }
      
      // Language-specific concerns
      const langConcerns = this.getLanguageSpecificConcerns(analysis.primaryLanguage);
      if (langConcerns.length > 0) {
        parts.push('');
        parts.push(`${this.getLanguageName(analysis.primaryLanguage)}-specific considerations:`);
        langConcerns.forEach(concern => parts.push(`- ${concern}`));
      }
      
      // Always include these core concerns
      parts.push('');
      parts.push('Core review points:');
      parts.push('- Code correctness and logic errors');
      parts.push('- Security vulnerabilities');
      parts.push('- Performance implications');
      parts.push('- Code style and best practices');
    }
    
    // Add output format hint for JSON
    if (options.format === 'json') {
      parts.push('');
      parts.push('Provide structured JSON output with: summary, issues[], suggestions[]');
    }
    
    return parts.join('\\n');
  }
  
  private static getLanguageSpecificConcerns(language: string): string[] {
    const concerns: Record<string, string[]> = {
      ts: [
        'Type safety and any usage',
        'Null/undefined handling',
        'Async/await error handling'
      ],
      tsx: [
        'React hooks dependencies',
        'Component re-rendering performance',
        'Props validation'
      ],
      js: [
        'Type coercion issues',
        'Async error handling',
        'ES6+ feature compatibility'
      ],
      py: [
        'Type hints accuracy',
        'Exception handling',
        'Python idioms and PEP compliance'
      ],
      go: [
        'Error handling patterns',
        'Goroutine safety',
        'Interface design'
      ],
      rs: [
        'Memory safety',
        'Error handling with Result<T,E>',
        'Lifetime correctness'
      ],
      java: [
        'Null pointer safety',
        'Thread safety',
        'Resource management'
      ]
    };
    
    return concerns[language] || [];
  }
  
  private static getLanguageName(ext: string): string {
    const names: Record<string, string> = {
      ts: 'TypeScript',
      tsx: 'React/TypeScript',
      js: 'JavaScript',
      jsx: 'React/JavaScript',
      py: 'Python',
      go: 'Go',
      rs: 'Rust',
      java: 'Java',
      cpp: 'C++',
      c: 'C'
    };
    return names[ext] || 'Programming language';
  }
  
  private static getPromptByName(promptName: string, analysis: PromptContext): string {
    const prompts: Record<string, string> = {
      'security': `Perform a security audit of these changes. Focus on:
- Authentication/authorization vulnerabilities
- Input validation and sanitization
- SQL injection, XSS, CSRF risks
- Sensitive data exposure
- Dependency vulnerabilities
- Cryptographic weaknesses
Provide specific vulnerabilities found with severity ratings.`,

      'performance': `Analyze performance implications of these changes:
- Time complexity of algorithms
- Memory usage and leaks
- Database query optimization
- Caching opportunities
- Network request optimization
- UI rendering performance
Suggest specific optimizations with expected impact.`,

      'refactor': `Review this refactoring for:
- Whether behavior is preserved
- Code clarity improvements
- SOLID principles adherence
- Reduced complexity metrics
- Test coverage maintenance
- Breaking change risks
Rate the refactoring success and suggest improvements.`,

      'feature': `Review this feature implementation:
- Completeness against requirements
- Edge cases handling
- Error scenarios
- User experience considerations
- Integration points
- Documentation needs
Identify any gaps in the implementation.`,

      'bugfix': `Verify this bug fix:
- Root cause properly addressed
- No regression risks
- Edge cases covered
- Tests added for the bug
- Related issues that might exist
Confirm the fix is complete and safe.`,

      'migration': `Review this migration/upgrade:
- Breaking changes identified
- Backwards compatibility
- Data migration safety
- Rollback procedures
- Performance impact
- Required documentation updates
Assess migration risk level.`,

      'api': `Review API changes for:
- Contract compatibility
- Versioning approach
- Error response formats
- Authentication/authorization
- Rate limiting needs
- Documentation completeness
Ensure API best practices are followed.`,

      'testing': `Evaluate test quality:
- Coverage of happy paths
- Edge case testing
- Error condition testing
- Test isolation and speed
- Mock/stub appropriateness
- Assertion quality
Suggest missing test scenarios.`
    };
    
    return prompts[promptName] || this.generateContextualPrompt(analysis, { format: 'markdown' });
  }
}