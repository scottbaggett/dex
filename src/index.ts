export { ContextEngine } from "./core/context";
export { GitExtractor } from "./core/git";
export { Formatter } from "./core/formatter";
export { MarkdownFormatter } from "./commands/extract/formatters/markdown";
export { JsonFormatter } from "./commands/extract/formatters/json";
export { XmlFormatter } from "./commands/extract/formatters/xml";

export type {
    DexOptions,
    OutputFormat,
    GitChange,
    ExtractedContext,
    SymbolMap,
    FormatterOptions,
} from "./types";
