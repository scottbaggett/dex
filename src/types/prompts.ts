export interface PromptTemplate {
    id: string;
    name: string;
    description: string;
    instructions: string;
    examples?: PromptExample[];
    tags?: string[];
    llm?: string[]; // Recommended LLMs for this preset
    extends?: string; // Inherit from another preset
    variables?: Record<string, string>; // Template variables
}

export interface PromptExample {
    input: string;
    output: string;
    explanation?: string;
}

export interface PromptsConfig {
    prompts: Record<string, Omit<PromptTemplate, "id">>;
    defaults?: {
        prompt?: string;
        llm?: string;
    };
}
