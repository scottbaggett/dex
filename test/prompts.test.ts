import { describe, it, expect } from 'vitest';
import { PromptGenerator } from '../src/core/prompts';
import { ExtractedContext, DexOptions } from '../src/types';

describe('PromptGenerator', () => {
  const createMockContext = (overrides: Partial<ExtractedContext> = {}): ExtractedContext => ({
    changes: [],
    scope: {
      filesChanged: 0,
      functionsModified: 0,
      linesAdded: 0,
      linesDeleted: 0
    },
    metadata: {
      generated: new Date().toISOString(),
      repository: { name: 'test-repo', branch: 'main', commit: 'abc123' },
      extraction: { depth: 'focused' },
      tokens: { estimated: 100 },
      tool: { name: 'dex', version: '0.1.0' }
    },
    ...overrides
  });

  describe('generate', () => {
    it('should use custom prompt when provided', () => {
      const context = createMockContext();
      const options: DexOptions = {
        prompt: 'Custom prompt for testing'
      };
      
      const result = PromptGenerator.generate(context, options);
      
      expect(result).toBe('Custom prompt for testing');
    });

    it('should generate contextual prompt for small changes', () => {
      const context = createMockContext({
        changes: [
          { file: 'test.js', status: 'modified', additions: 10, deletions: 5, diff: '' }
        ],
        scope: { filesChanged: 1, functionsModified: 1, linesAdded: 10, linesDeleted: 5 }
      });
      const options: DexOptions = {};
      
      const result = PromptGenerator.generate(context, options);
      
      expect(result).toMatchSnapshot();
      expect(result).toContain('Review this focused change for:');
    });

    it('should generate contextual prompt for large changes', () => {
      const context = createMockContext({
        changes: Array(10).fill(null).map((_, i) => ({
          file: `file${i}.js`,
          status: 'modified' as const,
          additions: 50,
          deletions: 30,
          diff: ''
        })),
        scope: { filesChanged: 10, functionsModified: 20, linesAdded: 500, linesDeleted: 300 }
      });
      const options: DexOptions = {};
      
      const result = PromptGenerator.generate(context, options);
      
      expect(result).toMatchSnapshot();
      expect(result).toContain('Review this substantial changeset for:');
    });

    it('should include task context when provided', () => {
      const context = createMockContext({
        task: { description: 'Fix authentication bug in login flow' }
      });
      const options: DexOptions = {};
      
      const result = PromptGenerator.generate(context, options);
      
      expect(result).toMatchSnapshot();
      expect(result).toContain('Fix authentication bug in login flow');
      expect(result).toContain('Whether the changes properly address the task');
    });

    it('should handle test file changes', () => {
      const context = createMockContext({
        changes: [
          { file: 'auth.test.js', status: 'modified', additions: 20, deletions: 5, diff: '' },
          { file: 'login.spec.ts', status: 'added', additions: 50, deletions: 0, diff: '' }
        ],
        scope: { filesChanged: 2, functionsModified: 5, linesAdded: 70, linesDeleted: 5 }
      });
      const options: DexOptions = {};
      
      const result = PromptGenerator.generate(context, options);
      
      expect(result).toMatchSnapshot();
      expect(result).toContain('Test coverage and quality');
    });

    it('should handle config file changes', () => {
      const context = createMockContext({
        changes: [
          { file: '.env.production', status: 'modified', additions: 3, deletions: 1, diff: '' },
          { file: 'config/database.js', status: 'modified', additions: 10, deletions: 5, diff: '' }
        ],
        scope: { filesChanged: 2, functionsModified: 0, linesAdded: 13, linesDeleted: 6 }
      });
      const options: DexOptions = {};
      
      const result = PromptGenerator.generate(context, options);
      
      expect(result).toMatchSnapshot();
      expect(result).toContain('Configuration changes and their impact');
    });

    it('should include language-specific concerns for TypeScript', () => {
      const context = createMockContext({
        changes: [
          { file: 'auth.ts', status: 'modified', additions: 30, deletions: 10, diff: '' },
          { file: 'types.d.ts', status: 'added', additions: 20, deletions: 0, diff: '' }
        ],
        scope: { filesChanged: 2, functionsModified: 3, linesAdded: 50, linesDeleted: 10 }
      });
      const options: DexOptions = {};
      
      const result = PromptGenerator.generate(context, options);
      
      expect(result).toMatchSnapshot();
      expect(result).toContain('TypeScript-specific considerations:');
      expect(result).toContain('Type safety and any usage');
    });

    it('should include language-specific concerns for React', () => {
      const context = createMockContext({
        changes: [
          { file: 'Button.tsx', status: 'modified', additions: 40, deletions: 20, diff: '' },
          { file: 'hooks/useAuth.tsx', status: 'added', additions: 60, deletions: 0, diff: '' }
        ],
        scope: { filesChanged: 2, functionsModified: 4, linesAdded: 100, linesDeleted: 20 }
      });
      const options: DexOptions = {};
      
      const result = PromptGenerator.generate(context, options);
      
      expect(result).toMatchSnapshot();
      expect(result).toContain('React/TypeScript-specific considerations:');
      expect(result).toContain('React hooks dependencies');
    });

    it('should add format hint for Grok', () => {
      const context = createMockContext();
      const options: DexOptions = { format: 'grok' };
      
      const result = PromptGenerator.generate(context, options);
      
      expect(result).toMatchSnapshot();
      expect(result).toContain('Provide structured JSON output');
    });

    it('should use prompt template when specified', () => {
      const context = createMockContext();
      const options: DexOptions = { promptTemplate: 'security' };
      
      const result = PromptGenerator.generate(context, options);
      
      expect(result).toMatchSnapshot();
      expect(result).toContain('security audit');
    });

    it('should handle mixed file types', () => {
      const context = createMockContext({
        changes: [
          { file: 'server.py', status: 'modified', additions: 25, deletions: 10, diff: '' },
          { file: 'client.js', status: 'modified', additions: 30, deletions: 15, diff: '' },
          { file: 'README.md', status: 'modified', additions: 5, deletions: 2, diff: '' }
        ],
        scope: { filesChanged: 3, functionsModified: 4, linesAdded: 60, linesDeleted: 27 }
      });
      const options: DexOptions = {};
      
      const result = PromptGenerator.generate(context, options);
      
      expect(result).toMatchSnapshot();
      expect(result).toContain('Documentation accuracy and completeness');
    });
  });

  describe('analyzeContext', () => {
    it('should correctly identify file types', () => {
      const context = createMockContext({
        changes: [
          { file: 'app.ts', status: 'modified', additions: 10, deletions: 5, diff: '' },
          { file: 'style.css', status: 'added', additions: 20, deletions: 0, diff: '' },
          { file: 'test.spec.js', status: 'modified', additions: 15, deletions: 8, diff: '' }
        ]
      });
      
      // This is a private method, but we can test it indirectly through generate
      const result = PromptGenerator.generate(context, {});
      
      expect(result).toContain('Test coverage and quality'); // Because we have test files
    });

    it('should detect primary language correctly', () => {
      const context = createMockContext({
        changes: [
          { file: 'main.go', status: 'modified', additions: 50, deletions: 20, diff: '' },
          { file: 'utils.go', status: 'added', additions: 30, deletions: 0, diff: '' },
          { file: 'config.json', status: 'modified', additions: 5, deletions: 2, diff: '' }
        ]
      });
      
      const result = PromptGenerator.generate(context, {});
      
      expect(result).toContain('Go-specific considerations:');
      expect(result).toContain('Error handling patterns');
    });
  });

  describe('Edge cases', () => {
    it('should handle empty context', () => {
      const context = createMockContext();
      const options: DexOptions = {};
      
      const result = PromptGenerator.generate(context, options);
      
      expect(result).toBeDefined();
      expect(result.length).toBeGreaterThan(0);
    });

    it('should handle context with no file extensions', () => {
      const context = createMockContext({
        changes: [
          { file: 'Makefile', status: 'modified', additions: 10, deletions: 5, diff: '' },
          { file: 'Dockerfile', status: 'added', additions: 20, deletions: 0, diff: '' }
        ]
      });
      
      const result = PromptGenerator.generate(context, {});
      
      expect(result).toBeDefined();
      expect(result).not.toContain('-specific considerations:'); // No language detected
    });

    it('should handle very long task descriptions', () => {
      const longTask = 'This is a very long task description '.repeat(20);
      const context = createMockContext({
        task: { description: longTask }
      });
      
      const result = PromptGenerator.generate(context, {});
      
      expect(result).toContain(longTask);
    });
  });
});