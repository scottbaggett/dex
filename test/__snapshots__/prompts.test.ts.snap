// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PromptGenerator > generate > should add format hint for Grok 1`] = `"Review this focused change for:\\n\\n\\nCore review points:\\n- Code correctness and logic errors\\n- Security vulnerabilities\\n- Performance implications\\n- Code style and best practices\\n\\nProvide structured JSON output with: summary, issues[], suggestions[]"`;

exports[`PromptGenerator > generate > should generate contextual prompt for large changes 1`] = `"Review this substantial changeset for:\\n\\n\\nJavaScript-specific considerations:\\n- Type coercion issues\\n- Async error handling\\n- ES6+ feature compatibility\\n\\nCore review points:\\n- Code correctness and logic errors\\n- Security vulnerabilities\\n- Performance implications\\n- Code style and best practices"`;

exports[`PromptGenerator > generate > should generate contextual prompt for small changes 1`] = `"Review this focused change for:\\n\\n- Test coverage and quality\\n- Whether tests properly validate the implementation\\n\\nJavaScript-specific considerations:\\n- Type coercion issues\\n- Async error handling\\n- ES6+ feature compatibility\\n\\nCore review points:\\n- Code correctness and logic errors\\n- Security vulnerabilities\\n- Performance implications\\n- Code style and best practices"`;

exports[`PromptGenerator > generate > should handle config file changes 1`] = `"Review this focused change for:\\n\\n- Configuration changes and their impact\\n- Environment compatibility\\n\\nJavaScript-specific considerations:\\n- Type coercion issues\\n- Async error handling\\n- ES6+ feature compatibility\\n\\nCore review points:\\n- Code correctness and logic errors\\n- Security vulnerabilities\\n- Performance implications\\n- Code style and best practices"`;

exports[`PromptGenerator > generate > should handle mixed file types 1`] = `"Review these changes for:\\n\\n- Documentation accuracy and completeness\\n- Whether docs match the implementation\\n\\nJavaScript-specific considerations:\\n- Type coercion issues\\n- Async error handling\\n- ES6+ feature compatibility\\n\\nCore review points:\\n- Code correctness and logic errors\\n- Security vulnerabilities\\n- Performance implications\\n- Code style and best practices"`;

exports[`PromptGenerator > generate > should handle test file changes 1`] = `"Review these changes for:\\n\\n- Test coverage and quality\\n- Whether tests properly validate the implementation\\n\\nTypeScript-specific considerations:\\n- Type safety and any usage\\n- Null/undefined handling\\n- Async/await error handling\\n\\nCore review points:\\n- Code correctness and logic errors\\n- Security vulnerabilities\\n- Performance implications\\n- Code style and best practices"`;

exports[`PromptGenerator > generate > should include language-specific concerns for React 1`] = `"Review these changes for:\\n\\n\\nReact/TypeScript-specific considerations:\\n- React hooks dependencies\\n- Component re-rendering performance\\n- Props validation\\n\\nCore review points:\\n- Code correctness and logic errors\\n- Security vulnerabilities\\n- Performance implications\\n- Code style and best practices"`;

exports[`PromptGenerator > generate > should include language-specific concerns for TypeScript 1`] = `"Review these changes for:\\n\\n\\nTypeScript-specific considerations:\\n- Type safety and any usage\\n- Null/undefined handling\\n- Async/await error handling\\n\\nCore review points:\\n- Code correctness and logic errors\\n- Security vulnerabilities\\n- Performance implications\\n- Code style and best practices"`;

exports[`PromptGenerator > generate > should include task context when provided 1`] = `"Review these changes in the context of: "Fix authentication bug in login flow"\\n\\nPlease analyze:\\n1. Whether the changes properly address the task\\n2. Any missing implementations or edge cases\\n3. Potential side effects or breaking changes"`;

exports[`PromptGenerator > generate > should use prompt template when specified 1`] = `
"Perform a thorough security audit of these code changes. Focus on:

1. **Authentication & Authorization**
   - Missing auth checks
   - Privilege escalation risks
   - Session management issues

2. **Input Validation**
   - Injection vulnerabilities (SQL, NoSQL, Command, LDAP)
   - XSS (stored, reflected, DOM-based)
   - XXE and deserialization flaws

3. **Data Protection**
   - Sensitive data exposure
   - Cryptographic weaknesses
   - Insecure data storage

4. **Security Misconfigurations**
   - Default credentials
   - Unnecessary features enabled
   - Verbose error messages

5. **Dependencies**
   - Known vulnerable packages
   - Outdated libraries

Provide findings in this format:
- **Severity**: Critical/High/Medium/Low
- **Location**: File and line number
- **Description**: Detailed explanation
- **Recommendation**: How to fix
"
`;
